gk/fib.c

```diff
@@ -1696,6 +1723,171 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
 	return del_fib_entry_numerical(&prefix_info, gk_conf);
 }
 
+/*
+ * Stack when function starts:
+ *
+ *   |            |
+ * 3 |   gk_conf  |
+ * 2 |    table   |
+ * 1 | prefix_str |
+ *   |____________|
+ */
+static int
+read_grantor_entries(lua_State *l,
+	struct ipaddr *gt_addrs, struct ipaddr *gw_addrs)
+{
+	lua_Integer i, tbl_size;
+
+	tbl_size = lua_objlen(l, 2);
+
+	/* Iterate over a table of tables. */
+	for (i = 1; i <= tbl_size; i++) {
+		const char *gt_ip, *gw_ip;
+		int ret;
+
+		/* Get the table at index i. */
+		lua_rawgeti(l, 2, i);
+
+		/*
+		 * Make sure that the item inside
+		 * the table is a table itself.
+		 */
+		if (!lua_istable(l, -1)) {
+			GK_LOG(ERR, "Grantor entry %ld is not a table\n", i);
+			return -1;
+		}
+
+		lua_getfield(l, 4, "gt_ip");
+		lua_getfield(l, 4, "gw_ip");
+
+		gt_ip = luaL_checkstring(l, 5);
+		gw_ip = luaL_checkstring(l, 6);
+
+		ret = convert_str_to_ip(gt_ip, &gt_addrs[i - 1]);
+		if (ret < 0) {
+			GK_LOG(ERR, "Cannot convert Grantor IP %s to bytes\n",
+				gt_ip);
+			return -1;
+		}
+
+		ret = convert_str_to_ip(gw_ip, &gw_addrs[i - 1]);
+		if (ret < 0) {
+			GK_LOG(ERR, "Cannot convert gateway IP %s to bytes",
+				gw_ip);
+			return -1;
+		}
+
+		/* Pop this Grantor/gateway IP pair and strings off stack. */
+		lua_pop(l, 3);
+	}
+
+	return 0;
+}
+
+static int
+add_grantor_entry_lb_verify_params(lua_State *l,
+	const char **prefix, lua_Integer *tbl_size,
+	struct gk_config **gk_conf)
+{
+	uint32_t ctypeid;
+	uint32_t correct_ctypeid_gk_config = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_GK_CONFIG_PTR);
+	void *cdata;
+	size_t len;
+
+	if (lua_gettop(l) != 3) {
+		GK_LOG(ERR,
+			"%s expected three arguments, however it received %d arguments\n",
+		      __func__, lua_gettop(l));
+		return -1;
+	}
+
+	/* First argument must be a prefix string. */
+	*prefix = lua_tolstring(l, 1, &len);
+	if (*prefix == NULL || len == 0) {
+		GK_LOG(ERR, "Invalid prefix for adding load balanced Grantor set\n");
+		return -1;
+	}
+
+	/* Second argument must be a table. */
+	luaL_checktype(l, 2, LUA_TTABLE);
+	*tbl_size = lua_objlen(l, 2);
+	if (*tbl_size <= 0) {
+		GK_LOG(ERR, "Table must have a positive number of Grantor entries\n");
+		return -1;
+	}
+
+	/* Third argument must be of type CTYPE_STRUCT_GK_CONFIG_PTR. */
+	cdata = luaL_checkcdata(l, 3,
+		&ctypeid, CTYPE_STRUCT_GK_CONFIG_PTR);
+	if (ctypeid != correct_ctypeid_gk_config) {
+		GK_LOG(ERR, "%s expected '%s' as the third argument\n",
+			__func__, CTYPE_STRUCT_GK_CONFIG_PTR);
+		return -1;
+	}
+	*gk_conf = *(struct gk_config **)cdata;
+
+	return 0;
+}
+
+int
+l_add_grantor_entry_lb(lua_State *l)
```

Comment:
All calls to `GK_LOG()` in this function and `add_grantor_entry_lb_verify_params()` and `read_grantor_entries()` should be replaced with `luaL_error()`, so the issue can be addressed in Lua. See similar functions for an example if needed.